eval.go

```diff
@@ -228,6 +229,11 @@ nextFileMatch:
 				continue
 			}
 
+			// Skip documents that don't belong to the tenant
+			if !tenant.EqualsID(ctx, repoMetadata.TenantID) {
+				continue
+			}
```

Comment:
Having the check at this level means shard merging should work out of the box. We can move it higher if we see it's a bottleneck.

Reply:
great, this is how I imagined how this would work originally. At the core loop we check, but then add some other logic in higher up places for efficiency reasons.

Reply:
Copying in my comment from the previous PR (https://github.com/sourcegraph/zoekt/pull/858). What's your new thinking on this?

For other tenant-aware services, we'd tried to design it so there's a "choke point" that clearly enforces tenancy. Often this is single interface and file guarded by `//🚨 SECURITY:` comments. Right now, we are only enforcing tenancy on `Search`, but not on `List`. And there are other places where we access index data, for example to get metadata or ngram stats.

I wonder if we could push this down further to where index data is loaded/ read. Like `reader.readTOC` and `reader.readIndexData`? That would help establish a single place on the read path where tenancy is enforced, as close to the data access as possible.

Reply:
From the perspective of indexer, tenant is treated just like any other metadata of the repo. It is similar to the tenant column in postgres. I think that is a good decision, because Zoekt doesn't need to know whether tenant is special.

For webserver, we enforce tenancy in `List` and in `Search` in the lowest level `Searcher`. `Searcher` is the core interface of Zoekt so it feels good to limit it that way. For example, in the code you can see it behaves just like a tombstone (see code adjacent to the tenant check indexdata).

We already trust these mechanics for private code, so it feels safe enough to use for tenants. WDYT?

Reply:
I chatted with @keegancsmith about this during our sync. We agreed my suggestion to use `reader.readIndexData` was nice, but that it would result in a much larger change since it'd also affect the indexing path (where you check if a repo already exists).

So I am +1 to go with this well-scoped change for now. We already plan to take a second look at multitenancy by switching to a push model for indexing. When we do this, it'd be great to take a second look here too and see if we can push the check down even further.

In the meantime, maybe we could document this design on `indexData` with our beloved `//🚨 SECURITY:` :)